zerver/tests/test_event_system.py

```diff
@@ -1211,7 +1211,7 @@ def test_queries(self) -> None:
         self.login_user(user)
 
         flush_per_request_caches()
-        with self.assert_database_query_count(37):
+        with self.assert_database_query_count(41):
```

Comment:
So the issue with the current implementation is that calls like `email_auth_enabled(realm)` now each do a db query. We can do this better by either:
1. Caching. But this solution I don't like too much because that adds complexity that perhaps we don't need? Then we;ll have to worry about when to delete the caches etc.
2. Just being smarter with how we do these db queries in sensitive codepaths (I think the event paths here might be the only ones) to just grab `realm.authentication_methods_dict()` (1 query) and take all the data needed from there instead of doing multiple queries calling `email_auth_enabled`, `password_auth_enabled`, etc. one-by-one there.